syntax = "proto2";
package tzrec.protos;

import "tzrec/protos/module.proto";
import "tzrec/protos/tower.proto";


message WideAndDeep {
    required MLP deep = 1;
    optional MLP final = 2;
    optional uint32 wide_embedding_dim = 3 [default = 4];
    // wide embedding init function, e.g. "nn.init.uniform_,a=-0.01,b=0.01"
    optional string wide_init_fn = 4;
}

message DeepFM {
    required MLP deep = 1;
    optional MLP final = 2;
    optional uint32 wide_embedding_dim = 3 [default = 4];
    // wide embedding init function, e.g. "nn.init.uniform_,a=-0.01,b=0.01"
    optional string wide_init_fn = 4;
}

message MultiTower {
    repeated Tower towers = 1;
    required MLP final = 2;
}

message MultiTowerDIN {
    repeated Tower towers = 1;
    repeated DINTower din_towers = 2;
    required MLP final = 3;
}

message MultiTowerDINTRT {
    repeated Tower towers = 1;
    repeated DINTower din_towers = 2;
    required MLP final = 3;
}


message DLRM {
  required MLP dense_mlp = 1;
  // whether to include sparse features after interaction
  optional bool arch_with_sparse = 2 [default=true];
  required MLP final = 3;
}


message MaskNet {
    required MaskNetModule mask_net_module = 1;
}
